File: 백준/Gold/15961. 회전 초밥/회전 초밥.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int d = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());
        int c = Integer.parseInt(st.nextToken());
        int[] count = new int[d + 1]; // 가짓수
        count[c] = 1; // 가짓수 저장
        int cnt = 1; // c 포함
        int[] arr = new int[N];
        for (int i = 0; i < N; i++) {
            arr[i] = Integer.parseInt(br.readLine());
            if (i < k) {
                count[arr[i]]++;
                if (count[arr[i]] == 1) cnt++;
            }
        }
        int ans = cnt;
        for (int i = k; i < N; i++) {
            count[arr[i - k]]--;
            if (count[arr[i - k]] == 0) cnt--;
            count[arr[i]]++;
            if (count[arr[i]] == 1) cnt++;
            ans = Math.max(ans, cnt);
        }
        for (int i = 0; i < k; i++) {
            count[arr[N + i - k]]--;
            if (count[arr[N + i - k]] == 0) cnt--;
            count[arr[i]]++;
            if (count[arr[i]] == 1) cnt++;
            ans = Math.max(ans, cnt);
        }
        System.out.println(ans);


    }
}

File: 백준/Gold/15961. 회전 초밥/회전 초밥.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.StringTokenizer
import kotlin.math.max


fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val (N, d, k, c) = readln().split(' ').map { it.toInt() }
    val count = IntArray(d + 1)
    count[c] = 1
    var cnt = 1 // c 포함
    val arr = IntArray(N)
    for (i in 0 until N) {
        arr[i] = readLine().toInt()
        if (i < k) {
            count[arr[i]]++
            if (count[arr[i]] == 1) cnt++
        }
    }
    var ans = cnt
    for (i in k until N) {
        count[arr[i - k]]--
        if (count[arr[i - k]] == 0) cnt--
        count[arr[i]]++
        if (count[arr[i]] == 1) cnt++
        ans = max(ans, cnt)
    }
    for (i in 0 until k) {
        count[arr[N + i - k]]--
        if (count[arr[N + i - k]] == 0) cnt--
        count[arr[i]]++
        if (count[arr[i]] == 1) cnt++
        ans = max(ans, cnt)
    }
    println(ans)
}